File: framework-cs/module/Devian/src/Core/VersionNumber.cs
// SSOT: skills/devian/10-module/20-core/14-version-number/SKILL.md

using System;

namespace Devian
{
    [Serializable]
    public struct VersionNumber : IComparable<VersionNumber>, IEquatable<VersionNumber>
    {
        public int Major;
        public int Minor;
        public int Patch;

        public VersionNumber(int major, int minor, int patch)
        {
            if (major < 0) throw new ArgumentOutOfRangeException(nameof(major), "Must be non-negative.");
            if (minor < 0) throw new ArgumentOutOfRangeException(nameof(minor), "Must be non-negative.");
            if (patch < 0) throw new ArgumentOutOfRangeException(nameof(patch), "Must be non-negative.");
            Major = major;
            Minor = minor;
            Patch = patch;
        }

        // ── Parse ──

        public static VersionNumber Parse(string value)
        {
            if (!TryParse(value, out var result))
                throw new FormatException($"Invalid version format: '{value}'. Expected '#.#.#'.");
            return result;
        }

        public static bool TryParse(string? value, out VersionNumber result)
        {
            result = default;
            if (string.IsNullOrEmpty(value)) return false;

            var parts = value.Split('.');
            if (parts.Length != 3) return false;

            if (!int.TryParse(parts[0], out int major) || major < 0) return false;
            if (!int.TryParse(parts[1], out int minor) || minor < 0) return false;
            if (!int.TryParse(parts[2], out int patch) || patch < 0) return false;

            result = new VersionNumber(major, minor, patch);
            return true;
        }

        // ── Comparison ──

        public int CompareTo(VersionNumber other)
        {
            int cmp = Major.CompareTo(other.Major);
            if (cmp != 0) return cmp;
            cmp = Minor.CompareTo(other.Minor);
            if (cmp != 0) return cmp;
            return Patch.CompareTo(other.Patch);
        }

        public static bool operator <(VersionNumber a, VersionNumber b) => a.CompareTo(b) < 0;
        public static bool operator >(VersionNumber a, VersionNumber b) => a.CompareTo(b) > 0;
        public static bool operator <=(VersionNumber a, VersionNumber b) => a.CompareTo(b) <= 0;
        public static bool operator >=(VersionNumber a, VersionNumber b) => a.CompareTo(b) >= 0;

        // ── Equality ──

        public bool Equals(VersionNumber other) =>
            Major == other.Major && Minor == other.Minor && Patch == other.Patch;

        public override bool Equals(object? obj) => obj is VersionNumber v && Equals(v);

        public override int GetHashCode()
        {
            unchecked
            {
                int hash = Major;
                hash = hash * 397 ^ Minor;
                hash = hash * 397 ^ Patch;
                return hash;
            }
        }

        public static bool operator ==(VersionNumber left, VersionNumber right) => left.Equals(right);
        public static bool operator !=(VersionNumber left, VersionNumber right) => !left.Equals(right);

        // ── Display ──

        public override string ToString() => $"{Major}.{Minor}.{Patch}";
    }
}

File: framework-ts/module/devian/src/versionNumber.ts
// SSOT: skills/devian/10-module/20-core/14-version-number/SKILL.md

// ============================================================================
// VersionNumber Type (#.#.# — Major.Minor.Patch)
// ============================================================================

export interface VersionNumber {
    readonly major: number;
    readonly minor: number;
    readonly patch: number;
}

// ============================================================================
// Construction
// ============================================================================

/**
 * Create a VersionNumber from major, minor, patch components.
 * All components must be non-negative integers.
 */
export function versionNumber(major: number, minor: number, patch: number): VersionNumber {
    if (!Number.isInteger(major) || major < 0)
        throw new Error(`major must be non-negative integer, got: ${major}`);
    if (!Number.isInteger(minor) || minor < 0)
        throw new Error(`minor must be non-negative integer, got: ${minor}`);
    if (!Number.isInteger(patch) || patch < 0)
        throw new Error(`patch must be non-negative integer, got: ${patch}`);
    return { major, minor, patch };
}

// ============================================================================
// Parse / TryParse
// ============================================================================

/**
 * Parse "#.#.#" format string into VersionNumber.
 * Throws on invalid input.
 */
export function parseVersionNumber(value: string): VersionNumber {
    const result = tryParseVersionNumber(value);
    if (result === null) {
        throw new Error(`Invalid version format: '${value}'. Expected '#.#.#'.`);
    }
    return result;
}

/**
 * Try parse "#.#.#" format string. Returns null on failure.
 */
export function tryParseVersionNumber(value: string): VersionNumber | null {
    if (!value) return null;

    const parts = value.split('.');
    if (parts.length !== 3) return null;

    const major = parseInt(parts[0], 10);
    const minor = parseInt(parts[1], 10);
    const patch = parseInt(parts[2], 10);

    if (isNaN(major) || major < 0) return null;
    if (isNaN(minor) || minor < 0) return null;
    if (isNaN(patch) || patch < 0) return null;

    return { major, minor, patch };
}

// ============================================================================
// Comparison
// ============================================================================

/**
 * Compare two VersionNumbers.
 * Returns negative if a < b, 0 if equal, positive if a > b.
 */
export function compareVersionNumber(a: VersionNumber, b: VersionNumber): number {
    if (a.major !== b.major) return a.major - b.major;
    if (a.minor !== b.minor) return a.minor - b.minor;
    return a.patch - b.patch;
}

/**
 * Check equality of two VersionNumbers.
 */
export function versionNumberEquals(a: VersionNumber, b: VersionNumber): boolean {
    return a.major === b.major && a.minor === b.minor && a.patch === b.patch;
}

// ============================================================================
// Display
// ============================================================================

/**
 * Format VersionNumber as "#.#.#" string.
 */
export function versionNumberToString(v: VersionNumber): string {
    return `${v.major}.${v.minor}.${v.patch}`;
}
